packages/yarnpkg-core/sources/Configuration.ts

```diff
@@ -376,6 +376,63 @@ export const coreDefinitions: {[coreSettingName: string]: SettingsDefinition} =
   },
 };
 
+export interface ConfigurationValueMap {
```

Comment:
Only worry I have (but still a massive improvement regardless) is that those types are separated from the configuration definition, so they could get out of sync if we're not careful.

Reply:
Yeah I understand the concern.
I was trying to come up with a solution where the type and the plugin's configuration properties are linked, but I haven't come up with a solution I like yet. I've got a couple of ideas left to try.

Reply:
I think I have a solution: https://www.typescriptlang.org/play?ts=4.0.2#code/CYUwxgNghgTiAEBbA9sArhBByAAgT1gDsAHAawHMB6MZOLeAbwCh54BLQgFxBgDMowCAMLJCvNuTQwonNqIBqUCGhABZKMUYtW8XsmQAueAGdOMDuQDc21gCNYRwmkS2e11gF8mXppzzEEABEQcUI2WVF4AF5tAB94AG1Tc0JyABpGPwCjAANkixyPAF04xKcXHgyGLJBc8tcYQqLrX38EAGUAsGjGeG0aMQkpGTlCI2YdRNIQPHZCeGm8ZF52RGJaTgAKXAIYEgpqWhAsAEoAOhFByWkIwkVlNQ0igH4jAFEADzMBTgAeYNC4VGGQSbDWG22+CIZCoNDo50u4muIwUShU6mICUWRQyUEIeCKAD4EgBGZraDzWSlMJgDUzwYjKcgcIydcA9CbwAZI4a3cY2XT6fmTVg1XL5VI5NICjzSyb2GDCkVi+B5MwFOU6LyeaXUphAA

Reply:
That doesn't work for subtypes, it wouldn't match 

```ts
interface ConfigurationValueMap {
  foo: '~' | '^' | '';
}
```

but I think I got it working!